source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2021 Mostafa Ashraf
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_changing_memory_semantics.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationChangingMemorySemantics::TransformationChangingMemorySemantics(
+    protobufs::TransformationChangingMemorySemantics message)
+    : message_(std::move(message)) {}
+
+TransformationChangingMemorySemantics::TransformationChangingMemorySemantics(
+    const protobufs::InstructionDescriptor& atomic_instruction,
+    uint32_t memory_semantics_operand_index,
+    uint32_t memory_semantics_new_value_id) {
+  *message_.mutable_atomic_instruction() = atomic_instruction;
+  message_.set_memory_semantics_operand_index(memory_semantics_operand_index);
+  message_.set_memory_semantics_new_value_id(memory_semantics_new_value_id);
+}
+
+bool TransformationChangingMemorySemantics::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  auto needed_atomic_instruction =
+      FindInstruction(message_.atomic_instruction(), ir_context);
+  // The atomic instruction must exist.
+  if (!needed_atomic_instruction) {
+    return false;
+  }
+
+  // Memory semantics operand Index must be equal to 2 or 3 only.
+  auto operand_index = message_.memory_semantics_operand_index();
+  if (operand_index != 2 && operand_index != 3) {
+    return false;
+  }
+
+  // The new instruction value must exist and must be OpConstant.
+  auto new_instruction = ir_context->get_def_use_mgr()->GetDef(
+      message_.memory_semantics_new_value_id());
```

Comment:
I would avoid calling this "new instruction". This sounds like we are adding a new instruction, but we are not. Maybe call it: `value_instruction`.

Reply:
Done.